File: src/inngest/functions/rfi-aging.ts
import { inngest } from "@/lib/inngest";
import { prisma } from "@/lib/db";

export const rfiAging = inngest.createFunction(
  { id: "rfi-aging" },
  { cron: "0 8 * * *" }, // Run daily at 8 AM
  async ({ step }) => {
    // Step 1: Find and flag overdue RFIs
    const overdueResults = await step.run("flag-overdue-rfis", async () => {
      const now = new Date();

      // Find RFIs that are past due but not yet flagged
      const overdueRfis = await prisma.rFI.findMany({
        where: {
          dueDate: { lt: now },
          isOverdue: false,
          status: { in: ["SUBMITTED", "PENDING_GC", "OPEN"] },
        },
        select: { id: true, projectId: true, rfiNumber: true, subject: true, createdById: true },
      });

      if (overdueRfis.length === 0) {
        return { flagged: 0, rfis: [] };
      }

      // Bulk update overdue flag
      await prisma.rFI.updateMany({
        where: { id: { in: overdueRfis.map((r) => r.id) } },
        data: { isOverdue: true },
      });

      return {
        flagged: overdueRfis.length,
        rfis: overdueRfis,
      };
    });

    // Step 2: Create overdue notifications
    if (overdueResults.flagged > 0) {
      await step.run("create-overdue-notifications", async () => {
        const notifications = overdueResults.rfis.map((rfi) => ({
          userId: rfi.createdById,
          type: "RFI_OVERDUE" as const,
          severity: "WARNING" as const,
          channel: "IN_APP" as const,
          title: `RFI ${rfi.rfiNumber} is overdue`,
          message: `"${rfi.subject}" has passed its response due date.`,
          projectId: rfi.projectId,
          entityId: rfi.id,
          entityType: "RFI",
          read: false,
        }));

        await prisma.notification.createMany({ data: notifications });

        return { notificationsCreated: notifications.length };
      });
    }

    // Step 3: Find RFIs approaching due date (within 2 days) and send reminders
    const approachingResults = await step.run("check-approaching-due", async () => {
      const now = new Date();
      const twoDaysFromNow = new Date(now.getTime() + 2 * 24 * 60 * 60 * 1000);

      const approachingRfis = await prisma.rFI.findMany({
        where: {
          dueDate: { gt: now, lte: twoDaysFromNow },
          isOverdue: false,
          status: { in: ["SUBMITTED", "PENDING_GC", "OPEN"] },
        },
        select: { id: true, projectId: true, rfiNumber: true, subject: true, dueDate: true, createdById: true },
      });

      if (approachingRfis.length === 0) {
        return { approaching: 0 };
      }

      // Check which ones already have a recent reminder (avoid duplicates)
      const recentReminders = await prisma.notification.findMany({
        where: {
          type: "RFI_RESPONSE_DUE",
          entityId: { in: approachingRfis.map((r) => r.id) },
          createdAt: { gt: new Date(now.getTime() - 24 * 60 * 60 * 1000) },
        },
        select: { entityId: true },
      });

      const alreadyNotified = new Set(recentReminders.map((n) => n.entityId));
      const needsReminder = approachingRfis.filter((r) => !alreadyNotified.has(r.id));

      if (needsReminder.length > 0) {
        const notifications = needsReminder.map((rfi) => ({
          userId: rfi.createdById,
          type: "RFI_RESPONSE_DUE" as const,
          severity: "INFO" as const,
          channel: "IN_APP" as const,
          title: `RFI ${rfi.rfiNumber} response due soon`,
          message: `"${rfi.subject}" is due ${rfi.dueDate!.toLocaleDateString("en-US", { month: "short", day: "numeric" })}.`,
          projectId: rfi.projectId,
          entityId: rfi.id,
          entityType: "RFI",
          read: false,
        }));

        await prisma.notification.createMany({ data: notifications });
      }

      return { approaching: needsReminder.length };
    });

    return {
      overdueFlagged: overdueResults.flagged,
      approachingReminders: approachingResults.approaching,
    };
  }
);

File: backend/app/tasks/rfi_aging.py
"""RFI aging Celery task.

Runs daily at 8 AM to flag overdue RFIs and send approaching-due reminders.
"""

import logging
from datetime import datetime, timedelta

from celery import shared_task
from sqlalchemy import select, update

from app.db.session import sync_session_factory
from app.models.enums import (
    NotificationChannel,
    NotificationSeverity,
    NotificationType,
    RFIStatus,
)
from app.models.notification import Notification
from app.models.rfi import RFI

logger = logging.getLogger(__name__)

# Statuses eligible for aging
AGING_STATUSES = [RFIStatus.SUBMITTED, RFIStatus.PENDING_GC, RFIStatus.OPEN]


@shared_task(name="rfi.aging", bind=True, max_retries=2)
def rfi_aging_check(self) -> dict:
    """Daily RFI aging: flag overdue, send approaching-due reminders."""
    try:
        return _run_aging()
    except Exception as exc:
        logger.exception("RFI aging check failed")
        raise self.retry(exc=exc, countdown=60)


def _run_aging() -> dict:
    """Execute the aging pipeline."""
    now = datetime.utcnow()
    overdue_flagged = 0
    approaching_reminders = 0

    with sync_session_factory() as session:
        # --- Step 1: Flag overdue RFIs ---
        result = session.execute(
            select(RFI).where(
                RFI.due_date < now,
                RFI.is_overdue == False,  # noqa: E712
                RFI.status.in_(AGING_STATUSES),
            )
        )
        overdue_rfis = result.scalars().all()

        for rfi in overdue_rfis:
            rfi.is_overdue = True

            # Create notification
            notification = Notification(
                user_id=rfi.created_by_id,
                type=NotificationType.RFI_OVERDUE,
                severity=NotificationSeverity.WARNING,
                channel=NotificationChannel.IN_APP,
                title=f"RFI {rfi.rfi_number} is overdue",
                message=f'"{rfi.subject}" has passed its response due date.',
                project_id=rfi.project_id,
                entity_id=rfi.id,
                entity_type="RFI",
            )
            session.add(notification)

        overdue_flagged = len(overdue_rfis)
        if overdue_flagged:
            logger.info("Flagged %d overdue RFIs", overdue_flagged)

        # --- Step 2: Approaching due (within 48 hours) ---
        cutoff = now + timedelta(hours=48)
        dedup_cutoff = now - timedelta(hours=24)

        result = session.execute(
            select(RFI).where(
                RFI.due_date > now,
                RFI.due_date <= cutoff,
                RFI.is_overdue == False,  # noqa: E712
                RFI.status.in_(AGING_STATUSES),
            )
        )
        approaching_rfis = result.scalars().all()

        for rfi in approaching_rfis:
            # Dedup: check if reminder already sent in last 24h
            existing = session.execute(
                select(Notification).where(
                    Notification.entity_id == rfi.id,
                    Notification.entity_type == "RFI",
                    Notification.type == NotificationType.RFI_RESPONSE_DUE,
                    Notification.created_at > dedup_cutoff,
                ).limit(1)
            )
            if existing.scalar_one_or_none():
                continue

            due_str = rfi.due_date.strftime("%b %d, %Y") if rfi.due_date else "soon"
            notification = Notification(
                user_id=rfi.created_by_id,
                type=NotificationType.RFI_RESPONSE_DUE,
                severity=NotificationSeverity.INFO,
                channel=NotificationChannel.IN_APP,
                title=f"RFI {rfi.rfi_number} response due soon",
                message=f'"{rfi.subject}" is due {due_str}.',
                project_id=rfi.project_id,
                entity_id=rfi.id,
                entity_type="RFI",
            )
            session.add(notification)
            approaching_reminders += 1

        if approaching_reminders:
            logger.info("Sent %d approaching-due reminders", approaching_reminders)

        session.commit()

    return {
        "overdueFlagged": overdue_flagged,
        "approachingReminders": approaching_reminders,
    }
